rasa/shared/core/events.py

```diff
@@ -507,17 +585,170 @@ def create_external(
         )
 
 
+# noinspection PyProtectedMember
+class DefinePrevUserUttered(Event):
+    """Defines the family of events that are used to update previous user utterance."""
+
+    def as_story_string(self) -> None:
+        """Returns the event as story string.
+
+        Returns:
+            None, as this event should not appear inside the story.
+        """
+        return
+
+
+# noinspection PyProtectedMember
+class DefinePrevUserUtteredFeaturization(DefinePrevUserUttered):
+    """Stores information whether action was predicted based on text or intent."""
+
+    type_name = "user_featurization"
+
+    def __init__(
+        self,
+        use_text_for_featurization: bool,
+        timestamp: Optional[float] = None,
+        metadata: Optional[Dict[Text, Any]] = None,
+    ) -> None:
+        """Creates event.
+
+        Args:
+            use_text_for_featurization: `True` if message text was used to predict
+                action. `False` if intent was used.
+            timestamp: When the event was created.
+            metadata: Additional event metadata.
+        """
+        self.use_text_for_featurization = use_text_for_featurization
+        super().__init__(timestamp, metadata)
+
+    def __str__(self) -> Text:
+        """Returns text representation of event."""
+        return f"DefinePrevUserUtteredFeaturization({self.use_text_for_featurization})"
+
+    def __hash__(self) -> int:
+        """Returns unique hash for event."""
+        return hash(self.use_text_for_featurization)
+
+    @classmethod
+    def _from_parameters(cls, parameters) -> "DefinePrevUserUtteredFeaturization":
+        return DefinePrevUserUtteredFeaturization(
+            parameters.get(USE_TEXT_FOR_FEATURIZATION),
+            parameters.get("timestamp"),
+            parameters.get("metadata"),
+        )
+
+    def as_dict(self) -> Dict[Text, Any]:
+        """Returns serialized event."""
+        d = super().as_dict()
+        d.update({USE_TEXT_FOR_FEATURIZATION: self.use_text_for_featurization})
+        return d
+
+    def apply_to(self, tracker: "DialogueStateTracker") -> None:
+        """Applies event to current conversation state.
+
+        Args:
+            tracker: The current conversation state.
+        """
+        if tracker.latest_action_name != ACTION_LISTEN_NAME:
+            # featurization belong only to the last user message
+            # a user message is always followed by action listen
+            return
+
+        # update previous user message's featurization based on this event
+        tracker.latest_message.use_text_for_featurization = (
+            self.use_text_for_featurization
+        )
+
+
+# noinspection PyProtectedMember
+class DefinePrevUserUtteredEntities(DefinePrevUserUttered):
```

Comment:
can we use something simpler like `EntityDetected`? should all entities labeled use this instead of being attached to the original user message?

Reply:
should the `UserUttered` be split into the actual user utterance and separate events for its interpretation (intent, entities)?

Reply:
we could, but it requires significant redesign of `tracker.events`